Exceptions during module initialization are swallowed

**Please complete the following information:**
- OS: FreeBSD, macOS
- Psycopg version: 2.8.6 (but still present in 2.9.6; see below)
- Python version: 3.11 (incompatible with 2.8.6; see below)
- PostgreSQL version: 16beta
- pip version: 23.1.2

**Describe the bug**

I accidentally installed psycopg2 version 2.8.6 on Python 3.11. This is not supported, but it took me a while to figure out what was wrong because of the following error message:
```
server/test_oauth.py:19: in <module>
    import psycopg2
../python_venv/lib/python3.11/site-packages/psycopg2/__init__.py:51: in <module>
    from psycopg2._psycopg import (                     # noqa
E   SystemError: initialization of _psycopg raised unreported exception
```

In this case, `initialization of _psycopg raised unreported exception` is generated when the module initialization function fails but returns a pointer anyway. It looks like Python expects the module to return `NULL` on failure instead. If I apply the following patch to 2.8.6:
```
diff --git a/psycopg/psycopgmodule.c b/psycopg/psycopgmodule.c
index 36fd1f9c..939875e1 100644
--- a/psycopg/psycopgmodule.c
+++ b/psycopg/psycopgmodule.c
@@ -1103,7 +1103,7 @@ INIT_MODULE(_psycopg)(void)
     if (!module) { goto exit; }
 
     if (0 > add_module_constants(module)) { goto exit; }
-    if (0 > add_module_types(module)) { goto exit; }
+    if (0 > add_module_types(module)) { return NULL; }
     if (0 > datetime_init()) { goto exit; }
     if (0 > mxdatetime_init(module)) { goto exit; }
     if (0 > encodings_init(module)) { goto exit; }
```
then I get a useful message:
```
server/test_oauth.py:19: in <module>
    import psycopg2
../../../build/python_venv/lib/python3.11/site-packages/psycopg2/__init__.py:51: in <module>
    from psycopg2._psycopg import (                     # noqa
E   SystemError: type psycopg2.extensions.ReplicationConnection has the Py_TPFLAGS_HAVE_GC flag but has no traverse function
```
which in this case pointed me to [the PR for 3.11 support](https://github.com/psycopg/psycopg2/pull/1376) and solved my problem.

Can the `INIT_MODULE` function be tweaked to always return `NULL` on failure? That would help with debugging corner cases; it looks like I'm not the only one to run into this [[1](https://bugzilla.redhat.com/show_bug.cgi?id=2021924), [2](https://github.com/psycopg/psycopg2/issues/1144), [3](https://github.com/psycopg/psycopg2/issues/995), [4](https://lists.mailman3.org/archives/list/mailman-users@mailman3.org/message/TKBQO3HF2R7AT6WB4R4V45QHHCWJ75QT/)].

--- a/NEWS
+++ b/NEWS
@@ -4,6 +4,8 @@ Current release
 What's new in psycopg 2.9.7 (unreleased)
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+- Fix propagation of exceptions raised during module initialization
+  (:ticket:`#1598`).
 - Fix building when pg_config returns an empty string (:ticket:`#1599`).
 - Wheel package compiled against OpenSSL 1.1.1v.
 

--- a/psycopg/psycopgmodule.c
+++ b/psycopg/psycopgmodule.c
@@ -1001,32 +1001,35 @@ INIT_MODULE(_psycopg)(void)
 
     /* initialize types and objects not exposed to the module */
     Py_SET_TYPE(&typecastType, &PyType_Type);
-    if (0 > PyType_Ready(&typecastType)) { goto exit; }
+    if (0 > PyType_Ready(&typecastType)) { goto error; }
 
     Py_SET_TYPE(&chunkType, &PyType_Type);
-    if (0 > PyType_Ready(&chunkType)) { goto exit; }
+    if (0 > PyType_Ready(&chunkType)) { goto error; }
 
     Py_SET_TYPE(&errorType, &PyType_Type);
     errorType.tp_base = (PyTypeObject *)PyExc_StandardError;
-    if (0 > PyType_Ready(&errorType)) { goto exit; }
+    if (0 > PyType_Ready(&errorType)) { goto error; }
 
-    if (!(psyco_null = Bytes_FromString("NULL"))) { goto exit; }
+    if (!(psyco_null = Bytes_FromString("NULL"))) { goto error; }
 
     /* initialize the module */
     module = PyModule_Create(&psycopgmodule);
-    if (!module) { goto exit; }
+    if (!module) { goto error; }
 
-    if (0 > add_module_constants(module)) { goto exit; }
-    if (0 > add_module_types(module)) { goto exit; }
-    if (0 > datetime_init()) { goto exit; }
-    if (0 > encodings_init(module)) { goto exit; }
-    if (0 > typecast_init(module)) { goto exit; }
-    if (0 > adapters_init(module)) { goto exit; }
-    if (0 > basic_errors_init(module)) { goto exit; }
-    if (0 > sqlstate_errors_init(module)) { goto exit; }
+    if (0 > add_module_constants(module)) { goto error; }
+    if (0 > add_module_types(module)) { goto error; }
+    if (0 > datetime_init()) { goto error; }
+    if (0 > encodings_init(module)) { goto error; }
+    if (0 > typecast_init(module)) { goto error; }
+    if (0 > adapters_init(module)) { goto error; }
+    if (0 > basic_errors_init(module)) { goto error; }
+    if (0 > sqlstate_errors_init(module)) { goto error; }
 
     Dprintf("psycopgmodule: module initialization complete");
-
-exit:
     return module;
+
+error:
+	if (module)
+		Py_DECREF(module);
+	return NULL;
 }

